clang/lib/Sema/SemaOpenACC.cpp

```diff
@@ -2589,6 +2589,59 @@ SemaOpenACC::ActOnOpenACCAsteriskSizeExpr(SourceLocation AsteriskLoc) {
   return BuildOpenACCAsteriskSizeExpr(AsteriskLoc);
 }
 
+/// Loops through a type and generates an appropriate InitListExpr to generate
+/// type initialization.
+static Expr *GenerateReductionInitRecipeExpr(ASTContext &Context,
+                                             SourceRange ExprRange,
+                                             QualType Ty) {
+  Ty = Ty.getCanonicalType();
+  llvm::SmallVector<Expr *> Exprs;
+
+  if (const RecordDecl *RD = Ty->getAsRecordDecl()) {
+    for (auto *F : RD->fields()) {
+      if (Expr *NewExpr =
+              GenerateReductionInitRecipeExpr(Context, ExprRange, F->getType()))
+        Exprs.push_back(NewExpr);
+      else
+        return nullptr;
+    }
+  } else if (const ConstantArrayType *AT = Context.getAsConstantArrayType(Ty)) {
```

Comment:
Can we ever get here with a non-constant array type? Is the type always constant because it's an initializer?

Reply:
`ConstantArrayType` means 'not a VLA' in this context (not that it is constant, it is that it is a constant size).

At the moment, we CAN get here with a pointer/VLA, but I haven't figured out how to implement that yet, and will probably do so next (for a few things).  I probably could use a NYI in the normal init generation for pointer types at the moment.  BUT since that is not Reduction specific, I'll do that in a separate patch if thats OK.

Reply:
There's code in `emitCXXNewAllocSize` (in CIRGenExprCXX.cpp in the incubator) to handle variable array length if you want to reuse that.

Reply:
Ah, well... its more than that.  The length of the VLA component is going to come in from a different place, via a variable in the recipe, so we have to figure out how to get that, what to loop over, etc.  It is more than just emitting the alloca, we have to do a full alloca of the right sizes then do the init.  But that is a good thing to know about for when I have to figure that out.